executable-spec/test/Cardano/Ledger/Spec/STS/Chain/Chain/Properties.hs

```diff
@@ -46,3 +53,243 @@ qc_revealsAreClassified
+updatePayloadPct
+  :: Trace.Trace (CHAIN ShortHash)
+  -> Float
+updatePayloadPct tr =
+  let -- get the total of transactions in tr
+      txTot = foldl' (\tot b -> tot + txsInAblock b) 0 blocks
+      blocks = (Trace.traceSignals Trace.NewestFirst tr)
+      txsInAblock = \b -> length $ Body.transactions (Chain.body b)
+      -- get the total of Txs with a non-empty update payload
+      txupdTot = foldl' (\tot b -> tot + updtxsInAblock b) 0 blocks
+      updtxsInAblock = \b -> length
+                               -- Txs with a non-empty update payload
+                               $ filter (\txb -> not . null
+                                                  $ Transaction.update txb
+                                        )
+                               $ map (Transaction.body)
+                               $ Body.transactions (Chain.body b)
+
+  in (fromIntegral txupdTot) / (fromIntegral txTot) * 100
+
+submittedSIPsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+submittedSIPsExist tr =
+  let lastSt = Trace.lastState tr
+  in Chain.subsips lastSt /= Map.empty
+
+revealedSIPsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+revealedSIPsExist tr =
+  let lastSt = Trace.lastState tr
+  in Chain.sipdb lastSt /= Map.empty
+
+sipBallotsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+sipBallotsExist tr =
+  let lastSt = Trace.lastState tr
+  in Chain.ballots lastSt /= Map.empty
+
+voteResultsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+voteResultsExist tr =
+  let lastSt = Trace.lastState tr
+      vResults = Map.elems $ Chain.vresips lastSt
+  in any (\vr -> Data.stakeInFavor vr /=  0
+                 && Data.stakeAgainst vr /= 0
+                 && Data.stakeAbstain vr /= 0
+         ) vResults
+
+-- Returns true if the last state of the inpuτ trace
+-- shows that all phased in the liefecycle of a software update
+-- have been covered
+lifecycleCoverage
+  :: Trace.Trace (CHAIN ShortHash)
+  -> Bool
+lifecycleCoverage tr =
+  let lastSt = Trace.lastState tr
+      subsips = Chain.subsips lastSt /= Map.empty
+      revsips = Chain.sipdb lastSt /= Map.empty
+      asips = Chain.asips lastSt /= Map.empty
+      ballots = Chain.ballots lastSt /= Map.empty
+      vresips = Chain.vresips lastSt /= Map.empty
+      apprvsips = Chain.apprvsips lastSt /= Set.empty
+  in subsips && revsips && asips && ballots && vresips && apprvsips
+
+-- Returns `True` if there is at least one SIP with a voting outcome
+-- as the one in the input parameter
+outcomeAny :: Data.TallyOutcome -> Trace.Trace (CHAIN ShortHash) -> Bool
+outcomeAny outc tr =
+  let lastSt = Trace.lastState tr
+      vresips = Chain.vresips lastSt
+      env = Trace._traceEnv tr
+      sDist = Chain.stakeDist env
+      pNoQ = Chain.prvNoQuorum env
+      pNoM = Chain.prvNoMajority env
+      r_a = Chain.r_a env
+  in any (\(_, outcome) -> outcome == outc )
+     $ Map.toList $ Data.tallyOutcomeMap vresips sDist pNoQ pNoM r_a
+
+stakeDistIsSkewed
```

Comment:
`stakeDistIsSkewed` and `statkeDistIsUniform` look quite similar, maybe we should factor out duplication?

Reply:
Yes sure.

Reply:
Done. OK.